language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -220,47 +242,86 @@ typeOf (TyInst x body ty) = do
+rewriteCtx ty@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
+    (st, _) <- get
+    case IM.lookup (unUnique u) st of
+        Just ty'@(NormalizedType TyVar{}) -> rewriteCtx (getNormalizedType ty')
+        Just ty'                          -> cloneType (getNormalizedType ty')
+        Nothing                           -> pure ty
+
 -- | Reduce any redexes inside a type.
-tyReduce :: Type TyNameWithKind a -> NormalizedType TyNameWithKind a
--- TODO: is this case actually safe? Don't we need to reduce again after substituting?
-tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = NormalizedType $ tySubstitute u (NormalizedType ty') (tyReduce ty) -- TODO: use the substitution monad here
-tyReduce (TyForall x tn k ty)                                                = NormalizedType $ TyForall x tn k (getNormalizedType $ tyReduce ty)
-tyReduce (TyFun x ty ty') | isTypeValue ty                                   = NormalizedType $ TyFun x (getNormalizedType $ tyReduce ty) (getNormalizedType $ tyReduce ty')
-                          | otherwise                                        = NormalizedType $ TyFun x (getNormalizedType $ tyReduce ty) ty'
-tyReduce (TyLam x tn k ty)                                                   = NormalizedType $ TyLam x tn k (getNormalizedType $ tyReduce ty)
-tyReduce (TyApp x ty ty') | isTypeValue ty                                   = NormalizedType $ TyApp x (getNormalizedType $ tyReduce ty) (getNormalizedType $ tyReduce ty')
-                          | otherwise                                        = NormalizedType $ TyApp x (getNormalizedType $ tyReduce ty) ty'
-tyReduce x                                                                   = NormalizedType x
+tyReduce :: Type TyNameWithKind () -> TypeCheckM a (NormalizedType TyNameWithKind ())
+tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = do
+    tyEnvAssign u (NormalizedType (void ty'))
+    tyReduce ty <* tyEnvDelete u
+tyReduce (TyForall x tn k ty)                                                = NormalizedType <$> (TyForall x tn k <$> (getNormalizedType <$> tyReduce ty))
+tyReduce (TyFun x ty ty') | isTypeValue ty                                   = NormalizedType <$> (TyFun x <$> (getNormalizedType <$> tyReduce ty) <*> (getNormalizedType <$> tyReduce ty'))
+                          | otherwise                                        = NormalizedType <$> (TyFun x <$> (getNormalizedType <$> tyReduce ty) <*> rewriteCtx ty')
+tyReduce (TyLam x tn k ty)                                                   = NormalizedType <$> (TyLam x tn k <$> (getNormalizedType <$> tyReduce ty))
+tyReduce (TyApp x ty ty') = do
+
+    let modTy = if isTypeValue ty
+        then fmap getNormalizedType . tyReduce
+        else rewriteCtx
+
+    tyRed <- getNormalizedType <$> tyReduce ty
+    let preTy = TyApp x tyRed <$> modTy ty'
+
+    if isTyLam tyRed
+        then tyReduce =<< preTy
```

Comment:
This feels strange. You reduce `ty`, then check whether it's a `TyLam` to avoid infinite looping and if it is, then jump to the first clause several lines above. Why special case actual reduction there? Why not just inline it here?

Reply:
So what this does differently (as compared to the pattern-matched clause now on line 334) is perform a reduction on the left *before* checking whether it is a type-level lambda.

The situation is in Fig. 7 of the spec: type reduction is to be performed first on lambas that are applied to a type, and then via type reduction frames.

Reply:
The spec just gives a small-step operational semantics for types. This is nice for a specification, but from the implementation point of view it's just weird and increases the cyclomatic complexity of a type checking algorithm. We do not have type environments in the specification either. I do not think the spec should affect the actual production algorithm that much, those type frames and the small-step operational semantics there just to give people an idea of how stuff works. It doesn't mean you have to implement it this way in production. CC'ing @mchakravarty

Reply:
Wait, so does your proposed modification deviate from the spec? I do not understand what you are proposing to do here.

Reply:
I propose to use one of the standard algorithms.

The first:

1. normalize `x`
2. normalize `f`
3. check whether `f` is a lambda
3.1. if so, then add `x` to the environment and proceed
3.2. if not, then just return the normalized `f` applied to the normalized `x`

The second is an abstract machine for which those type frames are specified.

Having said this, I also realized that in the first clause  

```haskell
tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = do
    tyEnvAssign u (NormalizedType (void ty'))
    tyReduce ty <* tyEnvDelete u
```

is incorrect, because you essentially postulate that whenever there is an application of a type lambda to a type, that type is normalized, which is just wrong.

What I propose deviates from the spec no more than using environments instead of nice-for-specification-but-pointless-for-implementation substitutions which are there in the figure 7.

Reply:
Ah you are correct. I will update the type state.

Reply:
Do you mean something like https://github.com/input-output-hk/plutus/blob/plutus-core/demo-failure/language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs#L341 or am I interpreting that wrong? The above code fails the test suite.

Reply:
The `tyEnvAssign u (void ty'')` part is strange at least, because `ty''` is the body of a function (why do you use those unreadable primes instead of descriptive names?) and you assign that body to the `u` variable instead of assigning an actual argument to the `u` variable (i.e. `arg`).

Reply:
I've pushed a corrected branch, however, it still fails the test suite.

Reply:
Another problem you have is

```haskell
_ -> pure $ NormalizedType $ TyApp x (getNormalizedType arg) fun
```

Here you return an argument applied to a function, i.e. the order is reversed. Besides

```haskell
    let modTy = if isTypeValue ty -- FIXME: does this recurse right?
        then fmap getNormalizedType . tyReduce
        else rewriteCtx
```

is pointless, you should always normalize.

Try something like this (I haven't tried to type check the code):

```haskell
tyReduce (TyApp x fun arg) = do
    nFun <- getNormalizedType <$> tyReduce fun
    nArg <- getNormalizedType <$> tyReduce arg
    case nFun of
        (TyLam _ (TyNameWithKind (TyName (Name _ _ uniq))) _ body) -> do
            tyEnvAssign u (void nArg)
            tyReduce body <* tyEnvDelete u
     _ -> pure $ NormalizedType $ TyApp x nFun nArg
```

If that doesn't work, I'll take a closer look.

Reply:
That fixes two of the errors, but unfortunately one remains. I have pushed a new branch in case you would like to examine the error.

Reply:
Yeah, I do.

Reply:
I've just checked. If you simply copy-paste what I wrote above and replace the existing `tyReduce (TyApp x fun arg)` clause, tests pass.